### ISSUE TITLE
Solution for adding authenticated Proxies (...almost!)

### ISSUE THREAD

**Issue by ENUM1** [ultrafunkamsterdam/undetected-chromedriver#498]
I am looking for a solution to use authenticated proxies with UC. The following allows me to insert authenticated proxies by inserting a chrome extension. However, it only works for `webdriver` and not UC.

The alternatives are to use `seleniumwire.undetected_chromedriver` but the handling of their TLS fingerprinting makes it unsuitable. So, I'm hoping for a crowd solution to this. 

All help is appreciated. 

```py
import os
import zipfile
import undetected_chromedriver.v2 as uc

PROXY_HOST = 'gb.smartproxy.com'  # rotating proxy
PROXY_PORT = 30000
PROXY_USER = 'user'
PROXY_PASS = 'pass'


manifest_json = """
{
    "version": "1.0.0",
    "manifest_version": 2,
    "name": "Chrome Proxy",
    "permissions": [
        "proxy",
        "tabs",
        "unlimitedStorage",
        "storage",
        "<all_urls>",
        "webRequest",
        "webRequestBlocking"
    ],
    "background": {
        "scripts": ["background.js"]
    },
    "minimum_chrome_version":"22.0.0"
}
"""

background_js = """
var config = {
        mode: "fixed_servers",
        rules: {
        singleProxy: {
            scheme: "http",
            host: "%s",
            port: parseInt(%s)
        },
        bypassList: ["localhost"]
        }
    };

chrome.proxy.settings.set({value: config, scope: "regular"}, function() {});

function callbackFn(details) {
    return {
        authCredentials: {
            username: "%s",
            password: "%s"
        }
    };
}

chrome.webRequest.onAuthRequired.addListener(
            callbackFn,
            {urls: ["<all_urls>"]},
            ['blocking']
);
""" % (
    PROXY_HOST,
    PROXY_PORT,
    PROXY_USER,
    PROXY_PASS,
)

pluginfile = "proxy_auth_plugin.zip"


def get_chromedriver(use_proxy=False, user_agent=None):
    path = os.path.dirname(os.path.abspath(__file__))
    chrome_options = uc.ChromeOptions()
    if use_proxy:
        pluginfile = 'proxy_auth_plugin.zip'

    with zipfile.ZipFile(pluginfile, "w") as zp:
        zp.writestr("manifest.json", manifest_json)
        zp.writestr("background.js", background_js)
    chrome_options = uc.ChromeOptions()
    chrome_options.add_extension(pluginfile)
    if user_agent:
        chrome_options.add_argument('--user-agent=%s' % user_agent)
    driver = uc.Chrome(options=chrome_options)
    return driver

def main():
    driver = get_chromedriver(use_proxy=True)
    #driver.get('https://www.google.com/search?q=my+ip+address')
    driver.get('https://httpbin.org/ip')


if __name__ == '__main__':
    main()
```

**Comment by taytanqbpass** [comment]
yep I actually also need this. But I'm getting issues with seleniumwire.undetected_chromedriver.v2 when packing the app with pyinstaller. i get some errors as discussed in https://github.com/wkeeling/selenium-wire/issues/469

**Comment by krypterro** [comment]
This authenticated proxy method works fine.

`def headless_driver(cbdb, node_data, vps_data, proxy_data, get_set):
        node_id = node_data["node_id"]
        vps_id = node_data["vps_id"]
        proxy_ip = proxy_data["proxy_ip"]
        proxy_user = proxy_data["proxy_user"]
        proxy_pass = proxy_data["proxy_pass"]
        proxy_port = proxy_data["proxy_port"]
        
        dir_path = os.path.dirname(os.path.realpath(__file__))
        
        binary_path = os.path.join(dir_path, "bin",
        "chromium-browser")
      
        chromedriver_path = os.path.join(dir_path, "bin", "chromedriver")
        
        ser = Service(chromedriver_path)

        options = webdriver.ChromeOptions()
        options.binary_location = binary_path
     
        options.add_argument('--headless')
        options.add_argument("--disable-gpu")
        	
        options.add_argument('--no-sandbox')
        options.add_argument("--disable-setuid-sandbox")
        dbug_port = random.randrange(20000,30000)
        options.add_argument(f"--remote-debugging-port={dbug_port}")

        options.add_argument("--disable-dev-shm-using")
        options.add_argument("--disable-extensions")
        
        options.add_argument("start-maximized")
        options.add_argument("disable-infobars")
        try:
            logger.info("creating driver")
            sw_options = {  # NOTE: Changed from 'options' to
                    "ignore_http_methods": [],  # 'sw_options' so no overwrite
                    "proxy": {
                        "http": "http://"
                                + proxy_user
                                + ":"
                                + proxy_pass
                                + "@"
                                + proxy_ip
                                + ":"
                                + str(proxy_port),
                                "https": "https://"
                                 + proxy_user
                                 + ":"
                                 + proxy_pass
                                 + "@"
                                 + proxy_ip
                                 + ":"
                                 + str(proxy_port),
                        "no_proxy": "localhost,127.0.0.1,dev_server:8080",
                    },
            }
  
            driver = uc.Chrome(
                        use_subprocess=True,
                        service=ser,
                        options=options,
                        seleniumwire_options=sw_options
            )

            brow_version = driver.capabilities['browserVersion']
            cdri_version = driver.capabilities['chrome']['chromedriverVersion'].split(' ')[0]
            logger.debug(f"browser: {brow_version}")
            logger.debug(f"driver: {cdri_version}")
            return driver

        except Exception as e:
            logger.error("Error in driver creation")
            error_msg = str(e)
            logger.error(e)
            status = "Crashed"
            crash_data = {}
            crash_data["node_id"] = node_id
            crash_data["vps_id"] = vps_id
            crash_data["error_msg"] = error_msg
            CB.insert_crash(cbdb, crash_data)`

**Comment by iamat8080** [comment]
option.add_argument(f"--load-extension={os.getcwd()}/Chrome/proxy")
